pyp2rpm/metadata_extractors.py

```diff
@@ -3,25 +3,89 @@
 import tempfile
 import shutil
 import itertools
+import glob
+from abc import ABCMeta
+try:
+    import xmlrpclib
+except ImportError:
+    import xmlrpc.client as xmlrpclib
 
 from pyp2rpm import archive
 from pyp2rpm import virtualenv
 from pyp2rpm.dependency_parser import deps_from_pyp_format, deps_from_pydit_json
 from pyp2rpm.exceptions import VirtualenvFailException
 from pyp2rpm.package_data import PackageData
+from pyp2rpm.logger import LoggerWriter
 from pyp2rpm import settings
 from pyp2rpm import utils
+from pyp2rpm import extract_distribution
 
 logger = logging.getLogger(__name__)
 
 
+def pypi_metadata_extension(extraction_fce):
+    """Extracts data from PyPI and appends them to data returned from 
+    given data extraction method.
+    """
+
+    def inner(self, client=None):
+        data = extraction_fce(self)
+        if client is None:
+            logger.debug("Client is None, skipping metadata extraction from PyPI.")
+            return data
+        try:
+            release_urls = client.release_urls(self.name, self.version)
+            release_data = client.release_data(self.name, self.version)
+        except:  # some kind of error with client => return TODO: log the failure
+            logger.debug('Client: {0} Name: {1} Version: {2}.'.format(
+                client, self.name, self.version))
+            logger.warn('Some kind of error while communicating with client: {0}.'.format(
+                client), exc_info=True)
+            failure_dict = {'md5' : 'FAILED TO EXTRACT FROM PYPI',
+                            'url' : 'FAILED TO EXTRACT FROM PYPI'}
+            data.set_from(failure_dict, update=True)
+            return data
+    
+        url = ''
+        md5_digest = None
+    
+        if len(release_urls):
```

Comment:
Do you mean the data update from failure_dict? Method set_from(dict, update=True) will append specific key-value data from dict dictionary to data object only if the key is not already there after previous extractions.